# Support Vector Machine (SVM) 
It is a powerful and versatile machine learning algorithm primarily used for classification tasks, but it can also be used for regression. SVM aims to find the optimal hyperplane that best separates the data points of different classes in the feature space.

### Key Concepts of SVM
Hyperplane:

In an n-dimensional space, a hyperplane is a flat affine subspace of dimension (n-1) that separates the space into two half-spaces. For a 2-dimensional space, the hyperplane is a line; for a 3-dimensional space, it is a plane.

#### Support Vectors:

Support vectors are the data points that are closest to the hyperplane and influence its position and orientation. These are the critical elements of the training set.
Margin:

The margin is the distance between the hyperplane and the nearest data points from either class. SVM aims to maximize this margin to achieve better generalization.
#### Kernel Trick:

The kernel trick allows SVM to handle non-linearly separable data by transforming the original feature space into a higher-dimensional space where a linear hyperplane can separate the classes. 
Common kernels include:
Linear Kernel: Suitable for linearly separable data.
Polynomial Kernel: Allows for curved decision boundaries.
Radial Basis Function (RBF) Kernel: Suitable for more complex, non-linear data.
Sigmoid Kernel: Similar to a neural network's activation function.
How SVM Works
#### Linear SVM:

For linearly separable data, SVM finds a hyperplane that separates the classes with the maximum margin. The hyperplane is defined by the equation 
𝑤⋅𝑥+𝑏=0
w⋅x+b=0, where 
𝑤
w is the weight vector and 
𝑏
b is the bias.

#### Non-Linear SVM:

For non-linearly separable data, SVM uses kernel functions to map the data into a higher-dimensional space where a linear hyperplane can separate the classes. This is known as the "kernel trick."

#### Soft Margin SVM:

SVM can handle cases where classes are not perfectly separable by introducing slack variables that allow some misclassification. The goal is to find a balance between maximizing the margin and minimizing classification errors. This is controlled by a regularization parameter 𝐶.


### Linear Kernel: 
Used when the data is linearly separable.
### Non-Linear Kernel: 
Used when the data is not linearly separable, with RBF being a common choice.
### Kernel Trick: 
Allows SVMs to perform classification in a higher-dimensional space without explicitly computing the transformation.

In [3]:
#import necessarey libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.svm import SVC


In [4]:
# Load the dataset
column_names = [
    'age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg',
    'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'
]
df=pd.read_csv(r'D:\Surya files\DataScience\ML\Classification\SVM\heartDiseaseDetection\input\processed.cleveland.data',header=None,names=column_names)

In [5]:
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  target  
0    3.0  0.0  6.0       0  
1    2.0  3.0  3.0       2  
2    2.0  2.0  7.0       1  
3    3.0  0.0  3.0       0  
4    1.0  0.0  3.0       0

In [7]:
print(f'{len(df.columns)} columns are there in tht given data set')

14 columns are there in tht given data set


The data you provided is from the Heart Disease UCI dataset, specifically the "processed.cleveland.data" file. Each row represents a patient's record with various medical attributes and the target variable indicating the presence of heart disease. Here’s a breakdown of what each column represents:

Age: Age of the patient in years.
Sex: Sex of the patient (1 = male; 0 = female).
cp: Chest pain type (1 = typical angina, 2 = atypical angina, 3 = non-anginal pain, 4 = asymptomatic).
trestbps: Resting blood pressure (in mm Hg on admission to the hospital).
chol: Serum cholesterol in mg/dl.
fbs: Fasting blood sugar > 120 mg/dl (1 = true; 0 = false).
restecg: Resting electrocardiographic results (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy).
thalach: Maximum heart rate achieved.
exang: Exercise induced angina (1 = yes; 0 = no).
oldpeak: ST depression induced by exercise relative to rest.
slope: The slope of the peak exercise ST segment (1 = upsloping, 2 = flat, 3 = downsloping).
ca: Number of major vessels (0-3) colored by fluoroscopy.
thal: Thalassemia (3 = normal; 6 = fixed defect; 7 = reversible defect).
target: Diagnosis of heart disease (0 = no heart disease, 1-4 = presence of heart disease).


### input:
63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0


### input explanation:
63.0: Age is 63 years.

1.0: Male.

1.0: Chest pain type is typical angina.

145.0: Resting blood pressure is 145 mm Hg.

233.0: Serum cholesterol is 233 mg/dl.

1.0: Fasting blood sugar > 120 mg/dl (true).

2.0: Resting electrocardiographic results show probable or definite left ventricular hypertrophy.

150.0: Maximum heart rate achieved is 150 bpm.

0.0: No exercise induced angina.

2.3: ST depression induced by exercise relative to rest is 2.3.

3.0: The slope of the peak exercise ST segment is downsloping.

0.0: No major vessels colored by fluoroscopy.

6.0: Thalassemia is fixed defect.

0: No heart disease (target variable).


